### PR TITLE
Fix quad light NAN on contact with other surfaces

### LINKED ISSUE #5912: Quad lights can produce abnormal values when intersecting a surface
**Describe the bug**

Quad lights can produce abnormal values when intersecting with a surface.

**Steps to reproduce**

1. Create a scene with a cube, a quad light, and a bloom post process.
2. Activate the bloom post process.
3. Move the quad light so that it collides with the cube.
4. Move the camera around the intersection point.
5. Observe the black squares.

**Expected behavior**

Light target is clean of any abnormal (`nan`, `INF`) values after the rendering pass when quad lights contacting a surface are present in the scene.

**Actual behavior**

Abnormal values are produced by the quad light into both the light and specular target.  
This behaviour is easily noticeable when used with a bloom postprocess where the `INF` value will get blown into sharp black squares.

**Screenshots/Video**

image

**Found in Branch**

`development`

**Desktop/Device (please complete the following information):**

 - Device: PC
 - OS: Windows
 - Version: 10
 - CPU: Intel i7-4790K
 - GPU: Nvidia GTX980
 - Memory: 4GB

**Additional context**

This bug is caused by the edge integration functions in `Ltc.azsli` where the inverse square root component will be `rsqrt(0)` under certain circumstances like described above.  
See: [IntegrateEdgeDiffuse#118](https://github.com/o3de/o3de/blob/development/Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli#L118)

That behaviour is prevented in Heitz's [original implementation](https://blog.selfshadow.com/sandbox/ltc.html) through the use of ternary check of theta before the division.

### PR DESCRIPTION
**Pull request by Bindless-Chicken**
- Quad lights can produce nan if intersecting other surfaces. This is caused by a division by 0 in the edge integration functions. Fixed by saturating the output.
- Small refactoring on LTC edge integration functions. To make the terms more explicit.

Fixes #5912 

Signed-off-by: Thomas Poulet <1039134+Bindless-Chicken@users.noreply.github.com>

### CODE REVIEW

**Comment by rgba16f** [comment]
Jenkins build & automated review process started.
https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-6174/

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli [hunk 1]
```diff
@@ -76,47 +76,43 @@ float3x3 BuildViewAlignedOrthonormalBasis(in float3 normal, in float3 dirToView)
...
+    // Cubic rational fitting of x/sin(x)
+    float a = 5.42031 + (3.12829 + 0.0902326 * absCosTheta) * absCosTheta;
+    float b = 3.45068 + (4.18814 + absCosTheta) * absCosTheta;
     float theta_sinTheta = a / b;
 
-    if (x < 0.0)
+    if (cosTheta < 0.0)
     {
-        theta_sinTheta = PI * rsqrt(saturate(1.0 - x * x)) - theta_sinTheta;
+        theta_sinTheta = PI * saturate(rsqrt(saturate(1.0 - cosTheta * cosTheta))) - theta_sinTheta;
```
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli:97]: good fix

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli [hunk 2]
```diff
@@ -76,47 +76,43 @@ float3x3 BuildViewAlignedOrthonormalBasis(in float3 normal, in float3 dirToView)
...
 }
 
 // Cheaper version of above which is good enough for diffuse
 float IntegrateEdgeDiffuse(float3 v1, float3 v2)
 {
     float cosTheta = dot(v1, v2);
     float absCosTheta = abs(cosTheta);
+
+    // Quadratic fitting of x/sin(x)
     float theta_sinTheta = 1.5708 + (-0.879406 + 0.308609 * absCosTheta) * absCosTheta;
```
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli:110]: nit: can we replace magic numbers and use defined constants for clarity?

**Review by Adi-Amazon — APPROVED** [review]
Requires testing on data for verification.  Perhaps @invertednormal can help with this

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli (hunk 1 above)
Comment (invertednormal) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli:97]: Hey I realize my suggestion from the bug wasn't correct here. The rsqrt of a 0-1 number will always be greater than 1, so the outer saturate is always going to clamp it to 1. I'd expect this to really throw off the results. I think we'll just have to eat the minor cost of a max() here.
```theta_sinTheta = PI * max(1.0, rsqrt(saturate(1.0 - cosTheta * cosTheta))) - theta_sinTheta;```

**Review by invertednormal — COMMENTED** [review]
Looks good - only concern I have is switching to always using the rational fit approach instead of acos(). It's been a while since I wrote this code but I'm think acos() may have been slightly faster, it just had accuracy issues on some platforms. It's likely a small performance difference either way.

Make sure to run the area light test in AtomSampleViewer to check for any changes in output. See https://o3de.org/docs/atom-guide/atom-sample-viewer/ for set up instructions. Once you have ASV running, use the menu to run a script, then select AreaLightTest.bv.luac. If there are any screenshot differences big enough to cause an error we'll want to look into why.

**Review by invertednormal — CHANGES_REQUESTED** [review]
(forgot to set my original review to request changes)

**Comment by invertednormal** [comment]
Also helpful for setting up the AtomSampleViewer project - https://o3de.org/docs/atom-guide/setup-atom-projects/

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli (hunk 2 above)
Comment (invertednormal) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli:110]: Not sure if there's much that can be done to make these values clearer. They're just the magic values for the quadratic approximation of x/sin(x)

Comment (Bindless-Chicken) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli:110]: Although I would agree on the form, I'd rather leave it that way. As @invertednormal pointed out these are just the terms of the formula determined by computer fitting, I would also argue that these shouldn't be left open for "customization" and that adding the comment on the formula might bring enough clarification. What do you think?

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli (hunk 1 above)
Comment (Bindless-Chicken) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli:97]: **On why I turned on the rational fitting by default.**
The rational fitting approach is the one that was recommended in the paper to address precision issues caused by acos not being a proper transcendental operation but just a proxy for code gen. (I also confirmed that the implementation given in the paper is still used).  
The reason I decided to activate that code path, is to unify the behaviour. That way we have an operation that will execute with similar precision on every targets, minimizing the chance of visual difference between targets.
On the performance side, I could not observe any meaningful difference, but the fitting approach uses one less VGPR.

**On the saturate.**
Good catch! I ran the tests and couldn't spot any difference between with and without. I'll rework that part and update the review.

**Review by invertednormal — APPROVED** [review]

**Pull request merged**